receiver/elasticsearchreceiver/config.go

```diff
@@ -0,0 +1,100 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package elasticsearchreceiver
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+	"go.uber.org/multierr"
+)
+
+// Config is the configuration for the elasticsearch receiver
+type Config struct {
+	scraperhelper.ScraperControllerSettings `mapstructure:",squash"`
+	confighttp.HTTPClientSettings           `mapstructure:",squash"`
+	// Username is the username used when making REST calls to elasticsearch. Must be specified if Password is. Not required.
+	Username string `mapstructure:"username"`
+	// Password is the password used when making REST calls to elasticsearch. Must be specified if Username is. Not required.
+	Password string `mapstructure:"password"`
+}
+
+var (
+	defaultEndpoint = "http://localhost:9200"
+)
+
+var (
+	errEndpointBadScheme    = errors.New("endpoint scheme must be http or https")
+	errUsernameNotSpecified = errors.New("password was specified, but not username")
+	errPasswordNotSpecified = errors.New("username was specified, but not password")
+	errEmptyEndpoint        = errors.New("endpoint must be specified")
+)
+
+var validSchemes = []string{
+	"http",
+	"https",
+}
+
+// Validate validates the given config, returning an error specifying any issues with the config.
+func (cfg *Config) Validate() error {
+	var combinedErr error
+	if err := invalidCredentials(cfg.Username, cfg.Password); err != nil {
+		combinedErr = multierr.Append(combinedErr, err)
+	}
+
+	if cfg.Endpoint == "" {
+		return multierr.Append(combinedErr, errEmptyEndpoint)
+	}
+
+	u, err := url.Parse(cfg.Endpoint)
+	if err != nil {
+		return multierr.Append(
+			combinedErr,
+			fmt.Errorf("invalid endpoint '%s': %w", cfg.Endpoint, err),
+		)
+	}
+
+	if !validScheme(u.Scheme) {
+		return multierr.Append(combinedErr, errEndpointBadScheme)
+	}
```

Comment:
The main reason for this check is that a URL without a scheme defined wonky. It's assumed that, without a scheme, the URL is actually an ABSOLUTE URL. See [the docs here](https://pkg.go.dev/net/url#Parse), specifically: `Trying to parse a hostname and path without a scheme is invalid but may not necessarily return an error, due to parsing ambiguities.`

I'd rather the config be considered invalid if not http/https, since those are the only supported protocols, but I'd be willing to replace this check with just `u.Scheme != ""`, if we think that's more appropriate.

Reply:
Ok, that's fine. Maybe we can do without the extra variables though. Something simple like this is sufficient:

```go
switch u.Scheme {
    case "http", "https": // ok
    default:
        multierr.Append(combinedErr, ...)
}
```